src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,899 @@
+        if (r < 0)
+               return r;
+
+        r = sd_netlink_message_append_u32(m, NFTA_SET_DATA_LEN, htobe32(dlen));
+        if (r < 0)
+                return r;
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+static int nft_add_element(sd_netlink *nfnl, sd_netlink_message **ret,
+                           int family, const char *set_name,
+                           const void *key, uint32_t klen,
+                           const void *data, uint32_t dlen) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        int r;
+
+        /*
+         * Ideally there would be an API that provides:
+         *
+         * 1) a init function to add the main ruleset skeleton
+         * 2) a function that populates the sets with all known address/port pairs to s/dnat for
+         * 3) a function that can remove address/port pairs again.
+         *
+         * At this time, the existing API is used which is built on a
+         * 'add/delete a rule' paradigm.
+         *
+         * This replicated here and each element gets added to the set
+         * one-by-one.
+         */
+        r = sd_nfnl_nft_message_new_setelems_begin(nfnl, &m, family, NFT_SYSTEMD_TABLE_NAME, set_name);
+        if (r < 0)
+                return r;
+
+        r = sd_nfnl_nft_message_add_setelem(m, 0, key, klen, data, dlen);
+        if (r < 0)
+                return r;
+
+        /* could theoretically append more set elements to add here */
+        r = sd_nfnl_nft_message_add_setelem_end(m);
+        if (r < 0)
+                return r;
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+static int nft_del_element(sd_netlink *nfnl,
+                           sd_netlink_message **ret, int family, const char *set_name,
+                           const void *key, uint32_t klen,
+                           const void *data, uint32_t dlen) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        int r;
+
+        r = sd_nfnl_nft_message_del_setelems_begin(nfnl, &m, family, NFT_SYSTEMD_TABLE_NAME, set_name);
+        if (r < 0)
+               return r;
+
+        r = sd_nfnl_nft_message_add_setelem(m, 0, key, klen, data, dlen);
+        if (r < 0)
+               return r;
+
+        r = sd_nfnl_nft_message_add_setelem_end(m);
+        if (r < 0)
+                return r;
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+/* This is needed so 'nft' userspace tool can properly format the contents
+ * of the set/map when someone lists their contents.
+ *
+ * The values cannot be changed, they are part of the nft tool type identifier ABI.
+ */
+#define TYPE_BITS 6
+
+enum nft_key_types {
+        TYPE_IPADDR = 7,
+        TYPE_IP6ADDR = 8,
+        TYPE_INET_PROTOCOL = 12,
+        TYPE_INET_SERVICE = 13,
+};
+
+static uint32_t concat_types2(enum nft_key_types a, enum nft_key_types b) {
+        return a << TYPE_BITS | b;
```

Comment:
hmm, i think it would be safer to cast `a` to `(uint32_t)` here first, to ensure that the value is wide enough for the `<<` to not overflow. C compilers have a lot of freedom how to map enums to numeric types after all.